keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -535,29 +599,109 @@ Disk-pressure eviction:
   - `reason=Evicted`
   - `message=The node was low on resource: ephemeral-storage. (...)`
 - Container status:
-  - `state=Ternminated`
+  - `state=Terminated`
   - `exitCode=137`
   - `reason=Error`
-- Retriable: Unclear, excessive disk usage suggests a bug or misconfiguration.
-  However, a restart on another node may succeed
 
-##### OOM kill
+##### Container memory limit exceeded
+
+Linux:
 
 - Reproduction: We run a job with a pod which attempts to allocate more
   memory than constrained in the container spec by `resources.limits.memory`
-- Comments: handled by kubelet
+- Comments: handled by kubelet in `kubelet/kubelet_pods.go` which merges-in the status update (setting of the reason field) by the container runtime
 - Pod status:
   - status: OOMKilled
   - `phase=Failed`
   - `reason=`
   - `message=`
 - Container status:
-  - `state=Ternminated`
+  - `state=Terminated`
   - `exitCode=137`
   - `reason=OOMKilled`
-- Retriable: Unclear, but if occurs when
-  `resources.requests.memory=resources.limits.memory` it strongly suggests
-  a software bug.
+
+Windows:
+
+- Reproduction: We run a job with a pod which attempts to allocate more
+  memory than constrained in the container spec by `resources.limits.memory`
+- Comments: there is not clear indication that the container failed due to exceeding memory limit
```

Comment:
We need to ensure this is properly documented in API that this is not yet functioning on Windows, same for graceful node shutdown.

Reply:
Limits are respected on Windows are enforced and containers are killed (tested). The issue is that we don't have a clear indication in the pod state of the reason. So, it should be probably documented in the API for the `ResourceExhausted` condition. 

I guess we can document it in comments for the `DisruptionTarget` and `ResourceExhausted` constants here: https://github.com/kubernetes/kubernetes/blob/de693b5e2d165081ef52dda83cd3aa3398f263b0/pkg/apis/core/types.go#L2436 during implementation.

Reply:
For my curiosity, can someone share the workload/container they used to test this on Windows?
I want to look at logs and see if we can get Windows to better report _why_ the container was killed.

Reply:
Sure, it is a workload for testing :)

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: test-script
data:
  eat.py: |
    import time, sys
    print("Hello world\n")

    CHUNK_SIZE = 1000000
    INIT_CHUNK_SIZE = 10000000
    init_chunk = ' ' * INIT_CHUNK_SIZE

    chunks = []
    allocated = 0
    chunks.append(init_chunk)
    allocated += len(init_chunk)
    for i in range(20000):
      chunk = str(i % 10) * CHUNK_SIZE
      chunks.append(chunk)
      allocated += len(chunk)

      print("Allocated: " + str(allocated) + ", waiting...")
      time.sleep(0.02)
---
apiVersion: batch/v1
kind: Job
metadata:
  name: job-oom-limit-win
spec:
  template:
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: kubernetes.io/os
                operator: In
                values:
                - windows
      volumes:
      - name: test-script
        configMap:
          name: test-script
      containers:
      - name: job-oom-limit
        image: winamd64/python
        command: ["python"]
        resources:
          limits:
            memory: "100Mi"
        args:
        - /test-script/eat.py
        imagePullPolicy: IfNotPresent
        volumeMounts:
          - mountPath: /test-script
            name: test-script
      restartPolicy: Never
  backoffLimit: 0
```